### ISSUE TITLE
VSTO Excel Addin TaskPane VB.Net CoreWebView2 Will Not Populate

### ISSUE THREAD

**Issue by TF523** [MicrosoftEdge/WebView2Feedback#622]
This issue doesn't seem to be resolved from #187. I am using vb.net to attempt to place WebView2 control into an excel vsto addin task pane and I am having no luck. I have used the EnsureCoreWebView2Async to attempt to get the core object to populate and there is simply nothing that will work. Converting the following lines to vb:

C#
var env = await CoreWebView2Environment.CreateAsync (null, "C:\temp");
await webView21.EnsureCoreWebView2Async (env);
webView21.CoreWebView2.Navigate ("https://bing.com")

VB.NET
Dim env as object
env = Await CoreWebView2Environment.CreateAsync(Nothing, "C:\Test")
Await WebView21.EnsureCoreWebView2Async(env)
WebView21.CoreWebView2.navigate("https://www.bing.com")

Still the core object will not assign.
I have attempted every conceivable combination which included using the CoreWebView2CreationProperties to set the path prior to the core object assignment, and still nothing has worked.

I am working off the Microsoft.Web.WebView2.1.0.674-prerelease.

I thought that it may be the browserexcutablefolder causing it to fail, but that doesn't seem to work when I tried set that either. I am at a loss...

[AB#30216412](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/30216412)

**Comment by champnic** [comment]
@TF523 Is it causing any error/exception? Also I'm not sure if it was just a copy error in your post, but you need to pass ENV to EnsureCoreWebView2Async(ENV). Do you have a runtime installed (Edge Canary/Dev/Beta, or WebView2 Runtime)?

**Comment by TF523** [comment]
yes it was just a copy error my last iteration I was trying other combination and that was left.

As far as the exception Initially when I tried to set the source it was giving me a write permission error just like they experienced on the other thread. When I moved to EnsureCoreWebView2Async that error went away and now the exception is object not assigned because it is nothing.

**Comment by TF523** [comment]
Thanks for following up so quickly...

**Comment by TF523** [comment]
Yes I have Dev version of runtime installed

**Comment by TF523** [comment]
I spun up a sample project to quickly test it in a vb.net form and everything works fine...

**Comment by champnic** [comment]
Do you have a repro app I could try out?

Another thing to try is to use the WebView2.CoreWebView2Ready event and see if that ever gets hit.

**Comment by TF523** [comment]
Unfortunately I cannot provide access to the repo it is part of a much larger proprietary project.

I dont see how the event is going to get triggered the object is nothing.

**Comment by champnic** [comment]
> Unfortunately I cannot provide access to the repo it is part of a much larger proprietary project.

Understandable.

Are you doing anything with threading, or is this all happening on your main UI thread?

I agree it looks like you are doing the right things here. I'll open a bug on our backlog and we'll try to repro and see what's going wrong.

**Comment by TF523** [comment]
No threading all on the Main UI.  All called in an Async sub.

**Comment by TF523** [comment]
Nic-
Can u send me what the default properties are for the core object.  Default execfolderpath and any other properties that may not be available, maybe we can set those... i was thinking that maybe due to a relative reference issue the object wont instantiate...
Thank you again for the help...
Let me know...-tyler

Sent from Yahoo Mail on Android 
 
  On Tue, Nov 10, 2020 at 10:16 PM, Nic Champagne Williamson [MSFT]<notifications@github.com> wrote:   



Unfortunately I cannot provide access to the repo it is part of a much larger proprietary project.


Understandable.

Are you doing anything with threading, or is this all happening on your main UI thread?

I agree it looks like you are doing the right things here. I'll open a bug on our backlog and we'll try to repro and see what's going wrong.

**Comment by champnic** [comment]
Thanks for the added context around how the control is being used.

The default when no exec folder is given is that the WebView2Loader.dll includes logic to see what versions of the runtime (Edge Canary/Dev/Beta or WebView2 Runtime) may be installed on the machine. It will find and use the most stable of those available. You have Dev installed, so it should be able to find and use that for any shipped version of our Nuget SDK package.

Some other things to double-check from #187:
1. Are you setting the "Source" property on WebView2 anywhere? If you are during initialization, you need to remove that to ensure that EnsureCoreWebView2Async is the call doing the initialization to pickup the correct UserDataFolder.
2. Is WebView2Loader.dll being deployed correctly when you run your project?

**Comment by TF523** [comment]
I am not using the source anywhere...
The .dll is in the deployment debug folder...

Sent from Yahoo Mail on Android 
 
  On Wed, Nov 11, 2020 at 11:56 AM, Nic Champagne Williamson [MSFT]<notifications@github.com> wrote:   


Thanks for the added context around how the control is being used.

The default when no exec folder is given is that the WebView2Loader.dll includes logic to see what versions of the runtime (Edge Canary/Dev/Beta or WebView2 Runtime) may be installed on the machine. It will find and use the most stable of those available. You have Dev installed, so it should be able to find and use that for any shipped version of our Nuget SDK package.

Some other things to double-check from #187:
   
   - Are you setting the "Source" property on WebView2 anywhere? If you are during initialization, you need to remove that to ensure that EnsureCoreWebView2Async is the call doing the initialization to pickup the correct UserDataFolder.
   - Is WebView2Loader.dll being deployed correctly when you run your project?

**Comment by TF523** [comment]
Have u attempted a simple vb.net vsto addin with this control placed on a taskpane??? Does it work 4 u???

Sent from Yahoo Mail on Android 
 
  On Wed, Nov 11, 2020 at 11:56 AM, Nic Champagne Williamson [MSFT]<notifications@github.com> wrote:   


Thanks for the added context around how the control is being used.

The default when no exec folder is given is that the WebView2Loader.dll includes logic to see what versions of the runtime (Edge Canary/Dev/Beta or WebView2 Runtime) may be installed on the machine. It will find and use the most stable of those available. You have Dev installed, so it should be able to find and use that for any shipped version of our Nuget SDK package.

Some other things to double-check from #187:
   
   - Are you setting the "Source" property on WebView2 anywhere? If you are during initialization, you need to remove that to ensure that EnsureCoreWebView2Async is the call doing the initialization to pickup the correct UserDataFolder.
   - Is WebView2Loader.dll being deployed correctly when you run your project?

**Comment by champnic** [comment]
No sorry I haven't had time to try this out myself. In general I don't have time to manually repro the issues people are running into, so I try to help as much as I can through debugging or asking for a reproducing app I can run on my machine. We have a bug on our backlog that would include one of our devs trying to reproduce this issue, but our backlog is large and I'm not sure when a dev will be able to try it out.

**Comment by TF523** [comment]
I will repo a simple build this afternoon for u to look at...

Sent from Yahoo Mail on Android 
 
  On Wed, Nov 11, 2020 at 12:10 PM, Nic Champagne Williamson [MSFT]<notifications@github.com> wrote:   


No sorry I haven't had time to try this out myself. In general I don't have time to manually repro the issues people are running into, so I try to help as much as I can through debugging or asking for a reproducing app I can run on my machine. We have a bug on our backlog that would include one of our devs trying to reproduce this issue, but our backlog is large and I'm not sure when a dev will be able to try it out.

**Comment by TF523** [comment]
Nic-
I have made a sample project for you. I have verified that the problem is solely with the core object and nothing to do with our code. To further investigate I have made both a TaskPane and a Form version of the WebView2 control.  What I have discovered is the Form version works fine, the TaskPane seems to be where the issue resides.  On the Form the display and navigation are flawless, on the TaskPane it immediately dies and gets a null object reference exception (just like what is happening on our larger project).  Please try the sample code to verify... please ignore my quick sloppy programming I am in a huge hurry trying to get this put together for you :-)
TF523/Excel_AddIn_WebView2


| 
| 
| 
|  |  |

 |

 |
| 
|  | 
TF523/Excel_AddIn_WebView2

This is a sample project that houses a Form and a TaskPane for the WeBview2 control to reside. For debugging of ...
 |

 |

 |


Thank you again for your help...
-Tyler


Confidentiality NoticeThis email may contain information that is privileged, confidential, or otherwise exempt from disclosure under applicable law. If you are not the addressee or it appears from the context that you have received this email in error, please advise me immediately by reply email, keep the contents confidential, and immediately delete the message and any attachments from your system. 

    On Wednesday, November 11, 2020, 12:10:36 PM PST, Nic Champagne Williamson [MSFT] <notifications@github.com> wrote:  
 
 


No sorry I haven't had time to try this out myself. In general I don't have time to manually repro the issues people are running into, so I try to help as much as I can through debugging or asking for a reproducing app I can run on my machine. We have a bug on our backlog that would include one of our devs trying to reproduce this issue, but our backlog is large and I'm not sure when a dev will be able to try it out.

**Comment by TF523** [comment]
Nic-
Verifying that the target user directory is working just fine when used with the form from within the vsto add-in.




Confidentiality NoticeThis email may contain information that is privileged, confidential, or otherwise exempt from disclosure under applicable law. If you are not the addressee or it appears from the context that you have received this email in error, please advise me immediately by reply email, keep the contents confidential, and immediately delete the message and any attachments from your system. 

    On Wednesday, November 11, 2020, 12:10:36 PM PST, Nic Champagne Williamson [MSFT] <notifications@github.com> wrote:  
 
 


No sorry I haven't had time to try this out myself. In general I don't have time to manually repro the issues people are running into, so I try to help as much as I can through debugging or asking for a reproducing app I can run on my machine. We have a bug on our backlog that would include one of our devs trying to reproduce this issue, but our backlog is large and I'm not sure when a dev will be able to try it out.

**Comment by TF523** [comment]
For your bug file.  The task pane WebView2 exception...




Confidentiality NoticeThis email may contain information that is privileged, confidential, or otherwise exempt from disclosure under applicable law. If you are not the addressee or it appears from the context that you have received this email in error, please advise me immediately by reply email, keep the contents confidential, and immediately delete the message and any attachments from your system. 

    On Wednesday, November 11, 2020, 12:10:36 PM PST, Nic Champagne Williamson [MSFT] <notifications@github.com> wrote:  
 
 


No sorry I haven't had time to try this out myself. In general I don't have time to manually repro the issues people are running into, so I try to help as much as I can through debugging or asking for a reproducing app I can run on my machine. We have a bug on our backlog that would include one of our devs trying to reproduce this issue, but our backlog is large and I'm not sure when a dev will be able to try it out.

**Comment by champnic** [comment]
@TF523 Thanks for putting together a sample, much appreciated! I'm not seeing it though - is there a link or attachment missing? (I'm using github.com - looks like you may be using email which could have dropped an attachement?)

**Comment by TF523** [comment]
https://github.com/tf523/Excel_AddIn_WebView2

**Comment by TF523** [comment]
Here is the exception thrown in debugger in case email broke the link again...

https://user-images.githubusercontent.com/5530663/98882435-1349a600-2441-11eb-9263-237758d0e3fb.PNG

**Comment by champnic** [comment]
Was able to repro and figured out the issue. Apparently we have a check for "DesignMode" that looks for whether or not we have a "Site", which is set on controls in the VisualStudio Designer. However, we assume that the presence of a "Site" means we are in designer, when in fact there are other circumstances that a Site is set - like the Task Pane. We are tracking a fix for this as part of #614, and will hopefully be available in the next release.

**Comment by TF523** [comment]
@champnic do u know when is the next update supposed to be released?

**Comment by champnic** [comment]
Probably around the end of this month.

**Comment by TF523** [comment]
Is there any access to the control.site property in runtime to override it as a workaround for now, or are we dead in the water until the update???

**Comment by TF523** [comment]
Nic-
Just downloaded the update and everything seems to be working great.  Thank you.

**Comment by TF523** [comment]
@champnic this may or may not be an issue, but I was setting up selenium to some automation testing on our embedded viewer (based on this article: https://docs.microsoft.com/en-us/microsoft-edge/webview2/howto/webdriver). Anyway I am having a tough time getting it to work as part of an add-in.  Also this article is a little dated as the selenium EdgeOptions constuctor has been updated as of their a-07 update.  I am probably just goobing the path executables for some reason as the add-in fires from a different location, but I was wondering if you know of anyone making this work in C# or vb.net as an add-in. Thank you...

**Comment by champnic** [comment]
Glad it's working now! I'm going to close this issue out, but if you see the same problem let me know.

Re: Selenium - We have an existing bug to update our docs - I just pinged the owner to see if we can update it sooner. Unfortunately I don't think I've seen or heard of anyone using selenium with add-ins yet.

**Comment by ShaunLoganOracle** [comment]
@TF523 , re:
> Just downloaded the update and everything seems to be working great. Thank you.
Which SDK and runtime bits are you using?
Thanks!

**Comment by TF523** [comment]
> @TF523 , re:
> > Just downloaded the update and everything seems to be working great. Thank you.
> 
> Which SDK and runtime bits are you using?
> Thanks!
> 

64 bit os visual studio .net 4.8

**Comment by ShaunLoganOracle** [comment]
Thanks.
More specifically, which [WebView2 SDK](https://docs.microsoft.com/en-us/microsoft-edge/webview2/releasenotes) and [WebView2 Runtime](https://developer.microsoft.com/microsoft-edge/webview2) (or [non-stable Microsoft Edge (Chromium) Canary channel](https://www.microsoftedgeinsider.com/download) ) are you using?

**Comment by TF523** [comment]
The latest pre-release webview2 from nuget. And latest edge developer release...

Sent from Yahoo Mail on Android 
 
  On Wed, Dec 2, 2020 at 6:51 AM, ShaunLoganOracle<notifications@github.com> wrote:   


Thanks.
More specifically, which WebView2 SDK and WebView2 Runtime (or non-stable Microsoft Edge (Chromium) Canary channel ) are you using?

**Comment by ShaunLoganOracle** [comment]
For others encountering this thread, in case it might help.
I ran into issues rendering content in a `UserControl` that contained a single WebView2 control, where the `UserControl` was added to the VSTO Task Pane in Excel.  More specifically, there were issues trying to set the `Source` property or call `CoreWebView2.NavigateToString()`. Those calls might be ignored or cause errors.
What I ended up doing is to defer those calls until my event handler for `CoreWebView2InitializationCompleted` got called.

**Env:**
C#, Winforms, .NET 4.8, Windows 10
SDK 1.721-prerelease
Canary/Edge 89.0.726.0
